### PR TITLE
[xivapi] Sort recipes by in game order.

### PR DESCRIPTION
**Pull request by garlond**
Order derived through experimentation.  See ApiRecipe::key for more detail.

### CODE REVIEW

File: xivapi/src/lib.rs [hunk 1]
```diff
@@ -103,6 +105,12 @@ struct RecipeLevelTable {
     SuggestedCraftsmanship: u32,
 }
 
+#[allow(non_snake_case)]
+#[derive(Clone, Debug, Deserialize)]
+pub struct GameContentLinks {
```
Comment (nvll) [inline xivapi/src/lib.rs]: Could you name this ApiGameContentLinks to be consistent with the others?

File: xivapi/src/lib.rs [hunk 2]
```diff
@@ -125,6 +133,78 @@ pub struct ApiRecipe {
     ItemIngredient3: Option<ItemIngredient>,
     ItemIngredient4: Option<ItemIngredient>,
     ItemIngredient5: Option<ItemIngredient>,
+    GameContentLinks: Option<GameContentLinks>,
+}
+
+impl ApiRecipe {
+    // XIV sorts is search results based on how they appear in the Crafting
```
Comment (nvll) [inline xivapi/src/lib.rs]: 'sorts its' ? Also, isn't "sorts search results by the crafting log" tautological when the crafting log is the search results?

File: xivapi/src/lib.rs [hunk 3]
```diff
@@ -125,6 +133,78 @@ pub struct ApiRecipe {
...
+    }
+}
+
+impl PartialEq for ApiRecipe {
+    fn eq(&self, other: &Self) -> bool {
+        self.ID == other.ID
+    }
+}
+
+impl Eq for ApiRecipe {}
```
Comment (nvll) [inline xivapi/src/lib.rs]: What is the empty Eq impl for?

File: xivapi/src/lib.rs [hunk 4]
```diff
@@ -215,4 +295,76 @@ mod test {
...
+        let api_results = query_recipe_by_name("gloves of aiming")?;
+
+        log::trace!("results: {:#?}", api_results);
+        assert_eq!(api_results.len(), names.len());
+        for (i, recipe) in api_results.iter().enumerate() {
+            assert_eq!(recipe.Name, names[i]);
+        }
+        Ok(())
+    }
+
```
Comment (nvll) [inline xivapi/src/lib.rs]: Could you add another test that confirms a Cloud Pearl for any job that isn't CRP is correct? It was made to work by the sort code you're replacing.

File: xivapi/src/lib.rs (hunk 1 above)
Comment (konkers) [reply xivapi/src/lib.rs]: None of the other sub structs are prefix with Api (CraftType, RecipeLevelTable, ItemIngredient).  Should I change those as well?

File: xivapi/src/lib.rs (hunk 2 above)
Comment (konkers) [reply xivapi/src/lib.rs]: Suggestions for a clearer explanation?

File: xivapi/src/lib.rs (hunk 3 above)
Comment (konkers) [reply xivapi/src/lib.rs]: Eq has no methods, rather declares that PartialEq also implements Eq: https://doc.rust-lang.org/std/cmp/trait.Eq.html

It looks like a better thing to do here is #[derive(Eq)] on the struct.

**Comment by konkers** [comment]
Now that I added the key() function to ApiRecipe, I'm going to remove the Ord impl and use sort_by_key()

File: xivapi/src/lib.rs (hunk 1 above)
Comment (nvll) [reply xivapi/src/lib.rs]: Oh whoops, good call. I forgot this one was embedded.

File: xivapi/src/lib.rs (hunk 2 above)
Comment (nvll) [reply xivapi/src/lib.rs]: I think you can remove the first sentence and start with "Through experimentation.." without losing anything.

File: xivapi/src/lib.rs (hunk 4 above)
Comment (nvll) [reply xivapi/src/lib.rs]: Don't worry about this, I should do it.

**Comment by konkers** [comment]
> Now that I added the key() function to ApiRecipe, I'm going to remove the Ord impl and use sort_by_key()

Scratch that. the
 ```        if let Some(ord) = self.partial_cmp(other) {
            ord
        } else {
            // Fall back on comparing ID if partial_cmp fails.
            self.CraftType.ID.cmp(&other.CraftType.ID)
        }```
makes doing it in place complicated.

File: xivapi/src/lib.rs (hunk 4 above)
Comment (konkers) [reply xivapi/src/lib.rs]: I'm happy to add it while I'm in here.

File: xivapi/src/lib.rs (hunk 3 above)
Comment (garlond) [reply xivapi/src/lib.rs]: looks like #[derive(Eq)] implements deep equality so this is still needed to follow the custom PartialEq relation ship setup here.

**Comment by garlond** [comment]
Should be good to go now

**Pull request merged**